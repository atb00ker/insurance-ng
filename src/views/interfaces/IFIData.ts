export interface IFIData {
  status: boolean;
  data: IFIUserData;
  error?: string;
}

export interface IFIUserData {
  name: string;
  date_of_birth: Date;
  pancard: string;
  ckyc_compliance: boolean;
  age_score: number;
  wealth_score: number;
  debt_score: number;
  investment_score: number;
  insurance: IFIInsurance[];
}

export interface IFIInsurance {
  uuid: string;
  type: string;
  title: string;
  description: string;
  account_id: string;
  score: number;
  current_premium: number;
  clauses: Array<string>;
  current_clauses: Array<string>;
  current_cover: number;
  offer_premium: number;
  offer_cover: number;
  is_active: boolean;
  yoy_deduction_rate: number;
  is_insuranceng_account: string;
}
