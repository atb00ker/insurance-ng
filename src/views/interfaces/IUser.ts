export interface IUser {
  id?: string;
  name?: string;
  email?: string;
  jwt: () => Promise<string>;
}

export interface IAuth {
  loginWithRedirect: (options?: any) => Promise<void>;
  logout: (options?: any) => void;
  isAuthenticated?: boolean;
  isReady?: boolean;
  user: IUser;
}
