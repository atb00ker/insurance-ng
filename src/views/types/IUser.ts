export type IUser = {
  id?: string;
  name?: string;
  email?: string;
  jwt: () => Promise<string>;
};

export type IAuth = {
  loginWithRedirect: (options?: any) => Promise<void>;
  logout: (options?: any) => void;
  isAuthenticated?: boolean;
  isReady?: boolean;
  user: IUser;
};

export type IUserProfileScores = {
  title: string;
  explaination: string;
  score: number;
};
