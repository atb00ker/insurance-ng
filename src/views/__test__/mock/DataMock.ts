import { InsuranceTypes } from '../../enums/Insurance';
import { IFIData, IFIInsurance } from '../../types/IFIData';
import { IUser } from '../../types/IUser';

export const userJohnDoe: IUser = {
  id: '12345',
  name: 'John Doe',
  email: 'john@example.com',
  jwt: () => new Promise(() => '12345'),
};

export const insuranceMedicalJohnDoe: IFIInsurance = {
  uuid: '94be9707-d419-404d-a3e2-5be54019fe85',
  type: InsuranceTypes.MEDICAL_PLAN,
  title: 'Medical Plan',
  description: 'Medical Plan Description',
  account_id: 'TESTACCT01',
  score: 0.5,
  clauses: ['Clause 1', 'Clause 2'],
  current_clauses: ['Clause 2', 'Clause 3'],
  current_premium: 1,
  current_cover: 10,
  offer_premium: 1.5,
  offer_cover: 11,
  is_active: true,
  is_claimed: false,
  yoy_deduction_rate: 0.5,
  is_insurance_ng_acct: false
}

export const fiDataJohnDoe: IFIData = {
  status: false,
  data: {
    name: 'John Doe',
    date_of_birth: new Date(2011,10,30),
    pancard: 'TESTPAN01',
    ckyc_compliance: true,
    age_score: 0,
    wealth_score: 0,
    debt_score: 0,
    investment_score: 0,
    phone: '999999999999999',
    shared_data_sources: 0,
    insurance: [insuranceMedicalJohnDoe]
  }
}
