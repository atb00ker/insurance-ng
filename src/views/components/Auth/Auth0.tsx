import React from 'react';
import { useAuth0, Auth0Provider, Auth0ProviderOptions } from '@auth0/auth0-react';
import { AuthContext } from './AuthProvider';
import { IAuth } from '../../types/IUser';

export const Provider = Auth0Provider;
export const ProviderOptions: Auth0ProviderOptions = {
  domain: process.env.AUTH0_DOMAIN || '',
  clientId: process.env.AUTH0_CLIENT_ID || '',
  redirectUri: process.env.AUTH0_REDIRECT_URI || window.location.origin,
  cacheLocation: 'localstorage',
};
const AuthConfigurations: React.FC = ({ children }) => {
  const { isLoading, isAuthenticated, loginWithRedirect, logout, user, getIdTokenClaims } = useAuth0();

  const auth0Logout = () => {
    return logout({ returnTo: process.env.AUTH0_LOGOUT_URI || window.location.origin });
  };

  const getJwt = async (): Promise<string> => {
    if (isAuthenticated) {
      const token = await getIdTokenClaims();
      return token.__raw;
    }
    return new Promise((_, __) => '');
  };

  const auth: IAuth = {
    loginWithRedirect: loginWithRedirect,
    logout: auth0Logout,
    isAuthenticated: isAuthenticated,
    isReady: !isLoading,
    user: {
      id: user?.sub,
      name: user?.name,
      email: user?.email,
      jwt: getJwt,
    },
  };

  // Enable if we want the autologin required feature
  // if (auth.isReady && !auth.isAuthenticated) {
  //   loginWithRedirect();
  //   return <div>Loading...</div>;
  // }
  return (
    <React.Fragment>
      <AuthContext.Provider value={auth}>{children}</AuthContext.Provider>
    </React.Fragment>
  );
};

export { AuthConfigurations };
