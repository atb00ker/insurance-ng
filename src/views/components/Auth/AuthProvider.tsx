import React from 'react';
import { IAuth } from '../../types/IUser';
import { AuthConfigurations, Provider, ProviderOptions } from './Auth0';

const AuthContext = React.createContext<IAuth>({} as IAuth);
const AuthProvider: React.FC = ({ children }) => {
  return (
    <Provider {...ProviderOptions}>
      <AuthConfigurations>{children}</AuthConfigurations>
    </Provider>
  );
};

export { AuthContext, AuthProvider };
