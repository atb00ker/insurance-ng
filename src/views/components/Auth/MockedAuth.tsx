import React from 'react';
import { AuthContext } from './AuthProvider';
import { IAuth, IUser } from '../../types/IUser';

const AuthProviderMock: React.FC<{
  children: React.ReactNode;
  isReady: boolean;
  isAuthenticated: boolean;
  user: IUser;
}> = ({ children, isReady, isAuthenticated, user }) => {
  const getJwt = async (): Promise<string> => {
    if (isAuthenticated) {
      return '12345';
    }
    return new Promise(() => '');
  };

  const auth: IAuth = {
    loginWithRedirect: () => {
      return new Promise(() => '');
    },
    logout: () => {},
    isAuthenticated: isAuthenticated,
    isReady: isReady,
    user: {
      id: user.id,
      name: user.name,
      email: user.email,
      jwt: getJwt,
    },
  };

  return <AuthContext.Provider value={auth}>{children}</AuthContext.Provider>;
};

export { AuthProviderMock };
