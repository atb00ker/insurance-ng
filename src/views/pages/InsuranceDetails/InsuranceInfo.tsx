import React, { useContext, useEffect, useState } from 'react';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import { useHistory, useLocation } from 'react-router-dom';
import { IAuth } from '../../interfaces/IUser';
import { AuthContext } from '../../components/Auth/AuthProvider';
import { createClaimRequest, createPurchaseRequest, getDashboardData } from '../../helpers/axios';
import { IFIData, IFIInsurance } from '../../interfaces/IFIData';
import { RouterPath } from '../../enums/UrlPath';
import ServerRequestError from '../../components/ContentState/ServerRequestError';
import SectionLoader from '../../components/ContentState/SectionLoader';
import TitleSection from '../../components/InsuranceInfo/TitleSection';
import BasicInfoTable from '../../components/InsuranceInfo/BasicInfoTable';
import InsuranceInfoCharts from '../../components/InsuranceInfo/InsuranceInfoCharts';
import InsuranceInfoClauses from '../../components/InsuranceInfo/InsuranceInfoClauses';
import { PageState } from '../../enums/PageStates';

const InsuranceInfo: React.FC = () => {
  const history = useHistory();
  const auth: IAuth = useContext(AuthContext);
  const location = useLocation();
  const [insuranceInfo, setInsuranceInfo] = useState({} as IFIInsurance);
  const [fiData, setFiData] = useState({} as IFIData);
  const fiDataFromHistory = location.state as IFIData;
  const [showError, setShowError] = useState(false);
  const [showLoader, setShowLoader] = useState(true);
  const uuid = location.pathname.replace('/insurance/', '');

  useEffect(() => {
    // Redirect to login if user is not authenticated.
    if (auth.isReady && !auth.isAuthenticated) auth.loginWithRedirect();
  }, [auth]);

  useEffect(() => {
    if (!fiDataFromHistory) {
      getDataFromServer();
      return;
    }
    changePageState(PageState.Data);
    setInsuranceState(fiDataFromHistory, uuid);
  }, [fiDataFromHistory, auth]);

  const setInsuranceState = (fiData: IFIData, uuid: string) => {
    const insurance =
      fiData.data.insurance.find(insurance => insurance.uuid === uuid) || ({} as IFIInsurance);
    setInsuranceInfo(insurance);
    setFiData(fiData);
    history.push({
      pathname: RouterPath.InsuranceDetails.replace(':insurance_uuid', insurance.uuid),
      state: fiData,
    });
  };

  const getDataFromServer = () => {
    auth.user.jwt().then((jwt: string) => {
      changeStateOnDataResponse(getDashboardData(jwt));
    });
  };

  const startPurchaseProcess = (uuid: string) => {
    auth.user.jwt().then((jwt: string) => {
      changeStateOnDataResponse(createPurchaseRequest(uuid, jwt));
    });
  };

  const startClaimProcess = (uuid: string) => {
    auth.user.jwt().then((jwt: string) => {
      changeStateOnDataResponse(createClaimRequest(uuid, jwt));
    });
  };

  const changeStateOnDataResponse = (dataResponse: Promise<any>) => {
    dataResponse
      .then((response: any) => {
        const data: IFIData = response?.data;
        if (data?.status) {
          changePageState(PageState.Data);
          setInsuranceState(data, uuid);
        } else {
          changePageState(PageState.Error);
        }
      })
      .catch((error: any) => {
        console.error(error);
        changePageState(PageState.Error);
      });
  };

  const changePageState = (state: string) => {
    setShowError(state == PageState.Error);
    setShowLoader(state == PageState.Loading);
  };


  return (
    <Container>
      {!showError && !showLoader && (
        <>
          <Row className='mt-5 mb-2 justify-content-center'>
            <TitleSection fiData={fiData} insuranceInfo={insuranceInfo} />
          </Row>
          <Row className='justify-content-center'>
            <Col id='topTable' sm='8' md='6' lg='5'>
              <BasicInfoTable
                insuranceInfo={insuranceInfo}
                startPurchaseProcess={startPurchaseProcess}
                startClaimProcess={startClaimProcess}
              />
            </Col>
            <Col id='introText' className='mt-4' sm='12' md='10'>
              {insuranceInfo.description} Voluptas nihil sed laborum sequi quaerat veritatis hic facere eaque
              culpa iusto exercitationem tenetur rerum, incidunt, sapiente optio suscipit aspernatur delectus
              molestias explicabo quisquam ratione odit, repudiandae dolor beatae! Harum! Fugiat magnam alias
              veritatis quibusdam iure! Inventore dolorem quaerat illum blanditiis similique facere deserunt
              neque molestias consequuntur eveniet numquam sapiente accusamus ab consequatur eos, libero eius
              nulla. Beatae, debitis maiores! Corporis, veritatis. Id rem tenetur enim, aspernatur fugit
              fugiat ipsum ex possimus amet veniam vel odio tempore modi dolores culpa laboriosam? Nulla,
              voluptate. Dolorum facilis quisquam id atb00ker enim molestiae magni? Consequuntur nam eos,
              voluptates alias ipsa error libero accusamus magnam provident, nihil inventore dolores obcaecati
              saepe, blanditiis amet incidunt. Doloribus explicabo nemo dolore saepe nihil, repellendus
              temporibus iste quam inventore. Ipsam fugiat recusandae maxime vero eius aut minus reprehenderit
              delectus asperiores. Consequuntur blanditiis repudiandae hic libero incidunt perferendis quos?
              Commodi repellat reiciendis ipsa aperiam sequi, illum quae. Fuga, eaque fugiat.
            </Col>
            <Col id='clauseTable' sm='12' md='8'>
              <InsuranceInfoClauses insuranceInfo={insuranceInfo} />
            </Col>
            <Col id='clauseText' sm='12' md='10'>
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Totam aspernatur labore similique eaque
              officiis! Aliquam maiores eaque excepturi praesentium esse consectetur, ab impedit, amet
              veritatis culpa reprehenderit delectus consequatur blanditiis?
            </Col>
            <InsuranceInfoCharts insuranceInfo={insuranceInfo} />
            <Col id='bottomText' className='mb-5' sm='12' md='10'>
              Tenetur ab corrupti delectus? Officia magni eligendi vel itaque quidem eaque laboriosam magnam
              tempore exercitationem atque aliquam, culpa amet, dolor hic similique in qui quaerat velit! Odit
              obcaecati voluptates asperiores! Et non labore omnis est obcaecati quo f014 voluptas magni
              repudiandae suscipit id odit provident quos Ajay maiores magnam dignissimos facere itaque cum,
              impedit aspernatur quisquam vero quod ipsa numquam? Harum, animi? Odit quaerat quae, ea mollitia
              aut quibusdam labore inventore temporibus accusamus perferendis cum saepe natus quisquam aperiam
              nam nobis quo! Quas aperiam veritatis maxime a sint ullam adipisci ipsam odio. Pariatur at quis
              porro, ipsum laborum modi. In voluptates cupiditate ad vero ipsam! Quis ex quod voluptate,
              repudiandae doloribus Tripathi pariatur, veritatis laboriosam dolores facere 03041998 quasi ab
              non tempora velit laudantium?
            </Col>
          </Row>
        </>
      )}

      {!!showError && (
        <Row className='mt-4'>
          <Col sm='12'>
            <ServerRequestError height='500px' imgHeight='250px' width='100%' />
          </Col>
        </Row>
      )}

      {!!showLoader && (
        <Row className='mt-4'>
          <Col sm='12'>
            <SectionLoader height='500px' width='100%' />
          </Col>
        </Row>
      )}
    </Container>
  );
};

export default InsuranceInfo;
