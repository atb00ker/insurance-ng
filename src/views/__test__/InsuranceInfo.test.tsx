/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom';
import 'regenerator-runtime/runtime';
import React from 'react';
import ReactDOM from 'react-dom';
import { render, cleanup, act, RenderResult } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { fiDataJohnDoe, insuranceMedicalJohnDoe, userJohnDoe } from './mock/DataMock';
import { IUser } from '../types/IUser';
import { AuthProviderMock } from '../components/Auth/MockedAuth';
import { TitleSection } from '../components/InsuranceInfo/TitleSection';
import { IFIData, IFIInsurance } from '../types/IFIData';

const titleSectionComponent = (isReady: boolean, isAuthenticated: boolean,
  fiData: IFIData = fiDataJohnDoe, insuranceInfo: IFIInsurance = insuranceMedicalJohnDoe,
  user: IUser = userJohnDoe) => (
  <AuthProviderMock isReady={isReady} isAuthenticated={isAuthenticated} user={user}>
    <Router>
      <TitleSection fiData={fiData} insuranceInfo={insuranceInfo} />
    </Router>
  </AuthProviderMock>
);

describe('NgFeatures page tests', () => {
  // beforeEach(() => {});
  afterEach(() => cleanup);

  it('renders without crashing', () => {
    const root = document.createElement('div');
    ReactDOM.render(titleSectionComponent(true, true), root);
  });

  it('test sign up button exists before auth', () => {
    let page: RenderResult | any;
    act(() => {
      page = render(titleSectionComponent(true, true));
    });
    expect(page.queryByText(insuranceMedicalJohnDoe.title)).toBeInTheDocument();
  });

  it('title section snapshot structure', () => {
    const snapTree = renderer.create(titleSectionComponent(true, true));
    expect(snapTree.toJSON()).toMatchSnapshot();
  });
});
