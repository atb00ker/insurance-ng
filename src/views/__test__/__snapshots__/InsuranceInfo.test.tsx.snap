// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgFeatures page tests title section snapshot structure 1`] = `
Array [
  <div
    className="col-sm-1"
  >
    <button
      className="btn-sm me-2 mb-2 buttons-height btn btn-outline-primary"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "minWidth": "80px",
        }
      }
      type="button"
    >
      <svg
        className="bi bi-arrow-return-left"
        fill="currentColor"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.5 1.5a.5.5 0 0 1 .5.5v4.8a2.5 2.5 0 0 1-2.5 2.5H2.707l3.347 3.346a.5.5 0 0 1-.708.708l-4.2-4.2a.5.5 0 0 1 0-.708l4-4a.5.5 0 1 1 .708.708L2.707 8.3H12.5A1.5 1.5 0 0 0 14 6.8V2a.5.5 0 0 1 .5-.5z"
          fillRule="evenodd"
        />
      </svg>
       
        
      Back
    </button>
  </div>,
  <div
    className="col-sm-10"
  >
    <h2
      className="text-center"
    >
      Medical Plan
    </h2>
  </div>,
  <div
    className="col-sm-1"
  />,
]
`;
